src/compiler/checker.ts

```diff
@@ -20042,6 +20042,12 @@ namespace ts {
             });
             if (!contextualMapper) {
                 inferTypes(context.inferences, getReturnTypeOfSignature(contextualSignature), getReturnTypeOfSignature(signature), InferencePriority.ReturnType);
+                const signaturePredicate = getTypePredicateOfSignature(signature);
+                const contextualPredicate = getTypePredicateOfSignature(sourceSignature);
+                if (signaturePredicate && contextualPredicate && signaturePredicate.kind === contextualPredicate.kind &&
+                    (signaturePredicate.kind !== TypePredicateKind.Identifier || signaturePredicate.parameterIndex === (contextualPredicate as IdentifierTypePredicate).parameterIndex)) {

```

Comment:
would probably be a little safer to check `signaturePredicateKind === TypePredicateKind.This ||`, in case we add more kinds in the future.

In the meantime, it's quite a bit easier to read.